### PR TITLE
Feature/50 support merge statement

### LINKED ISSUE #50: Support MERGE Statement
# Situation

The [MERGE](https://docs.exasol.com/sql/merge.htm) statement provides a combination of `INSERT`, `UPDATE` and `DELETE`.

It is especially useful in cases where new data should be combined with data that already exists in a table.

# Acceptance Criteria
1. `MERGE` supported without `WHEN`

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #50.

### CODE REVIEW

File: doc/user_guide/statements/merge.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    .merge("destination")
+    .using("source")
+    .on(eq(column("source", "id"), column("destination", "id");
+```
+
+### Fine Tuning the `MERGE` Strategies
+
+If you need more control over what happens in case of matching rows, you can add `whenMatched()`.
+
+In case of `whenMatched()` select one of the two strategies `thenInsert()` &hellip;
```
Comment (AnastasiiaSergienko) [inline doc/user_guide/statements/merge.md]: One of two `thenInsert()` strategies? Or are they different?

File: doc/user_guide/statements/merge.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+```java
+merge.whenMatched()
+    .thenUpdate() //
+    .setToDefault("c2") //
+    .set("c3", "foo") //
+    .set("c4", 42);
+```
+
+&hellip; or or `thenDelete()`.
```
Comment (AnastasiiaSergienko) [inline doc/user_guide/statements/merge.md]: Double or

File: doc/user_guide/statements/merge.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+### Fine Tuning the `MERGE` Strategies
+
+If you need more control over what happens in case of matching rows, you can add `whenMatched()`.
+
+In case of `whenMatched()` select one of the two strategies `thenInsert()` &hellip;
+
+```java
+merge.whenMatched()
+    .thenUpdate() //
```
Comment (AnastasiiaSergienko) [inline doc/user_guide/statements/merge.md]: I'm confused. It's said insert above, but update in the example

File: src/main/java/com/exasol/sql/Table.java [hunk 4]
```diff
@@ -2,15 +2,16 @@
...
+import com.exasol.sql.dml.merge.MergeVisitor;
+import com.exasol.sql.dql.select.SelectVisitor;
 
 /**
  * This class represents a {@link Table} in an SQL Statement
  */
 public class Table extends AbstractFragment {
     private final String name;
-    private final Optional<String> as;
+    private final String as;
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/sql/Table.java]: The name is suspicious

Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/sql/Table.java]: I had to find a comment to figure out what it means

File: doc/user_guide/statements/merge.md (hunk 2 above)
Comment (redcatbear) [reply doc/user_guide/statements/merge.md]: Removed.

File: src/main/java/com/exasol/sql/expression/rendering/ValueExpressionRenderer.java [hunk 5]
```diff
@@ -24,8 +28,23 @@ public void visit(final IntegerLiteral literal) {
         append(literal.toString());
     }
 
+    @Override
+    public void visit(final ColumnReference columnReference) {
+        final String tableName = columnReference.getTableName();
+        if ((tableName != null) && !tableName.isEmpty()) {
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/sql/expression/rendering/ValueExpressionRenderer.java]: Why tableName != null is in the brackets and !tableName.isEmpty() is not? It's somehow inconsistent.

File: doc/user_guide/statements/merge.md (hunk 3 above)
Comment (redcatbear) [reply doc/user_guide/statements/merge.md]: Fixed.

File: src/main/java/com/exasol/sql/Table.java (hunk 4 above)
Comment (redcatbear) [reply src/main/java/com/exasol/sql/Table.java]: Changed to `alias` for more clarity.

Comment (redcatbear) [reply src/main/java/com/exasol/sql/Table.java]: Also adapted the getter.

File: src/test/java/com/exasol/hamcrest/SqlFragmentRenderResultMatcher.java [hunk 6]
```diff
@@ -38,7 +40,7 @@ public static SqlFragmentRenderResultMatcher rendersTo(final String expectedText
     /**
      * Factory method for {@link SqlFragmentRenderResultMatcher}
```
Comment (AnastasiiaSergienko) [inline src/test/java/com/exasol/hamcrest/SqlFragmentRenderResultMatcher.java]:
```suggestion
     * Factory method for {@link SqlFragmentRenderResultMatcher}.
```

File: src/test/java/com/exasol/sql/ddl/create/TestCreateTable.java [hunk 7]
```diff
@@ -1,16 +1,25 @@
...
-
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.instanceOf;
 import static org.junit.jupiter.api.Assertions.assertAll;
 
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import com.exasol.datatype.type.Boolean;
```
Comment (AnastasiiaSergienko) [inline src/test/java/com/exasol/sql/ddl/create/TestCreateTable.java]: The same problem with imports

File: src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java [hunk 8]
```diff
@@ -24,13 +24,13 @@ void getTableName() {
 
     @Test
     void getIfExistsFalseByDefault() {
-        assertFalse(this.dropSchema.getIfExists());
+        assertFalse(this.dropSchema.hasIfExistsModifier());
```
Comment (AnastasiiaSergienko) [inline src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java]: I don't get what this method is doing after I read the name :(

File: src/test/java/com/exasol/sql/expression/ColumnReferenceTest.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
+package com.exasol.sql.expression;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+
+import org.junit.jupiter.api.Test;
+
+class ColumnReferenceTest {
+
```
Comment (AnastasiiaSergienko) [inline src/test/java/com/exasol/sql/expression/ColumnReferenceTest.java]: extra line

File: src/main/java/com/exasol/sql/ddl/create/rendering/CreateTableRenderer.java [hunk 10]
```diff
@@ -1,15 +1,29 @@
 package com.exasol.sql.ddl.create.rendering;
 
+import com.exasol.datatype.type.AbstractStringDataType;
 import com.exasol.datatype.type.Boolean;
-import com.exasol.datatype.type.*;
+import com.exasol.datatype.type.Char;
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/sql/ddl/create/rendering/CreateTableRenderer.java]: Me too. I ran "organize imports" on the whole project. Should be fixed now.

File: src/main/java/com/exasol/sql/dml/insert/AbstractInsertValueTable.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
+package com.exasol.sql.dml.insert;
+
+import com.exasol.sql.*;
+
+public abstract class AbstractInsertValueTable<T extends AbstractInsertValueTable<T>> extends AbstractFragment {
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/sql/dml/insert/AbstractInsertValueTable.java]: Added.

File: src/main/java/com/exasol/sql/dml/merge/MergeInsertClause.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
+package com.exasol.sql.dml.merge;
+
+import com.exasol.sql.Fragment;
+import com.exasol.sql.dml.insert.AbstractInsertValueTable;
+
+public class MergeInsertClause extends AbstractInsertValueTable<MergeInsertClause> implements MergeFragment {
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/sql/dml/merge/MergeInsertClause.java]: Added.

File: src/main/java/com/exasol/sql/dml/merge/MergeVisitor.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
+package com.exasol.sql.dml.merge;
+
+import com.exasol.sql.*;
+import com.exasol.sql.dml.insert.InsertFields;
+
+public interface MergeVisitor extends TableValuesVisitor {
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/sql/dml/merge/MergeVisitor.java]: Added.

File: src/main/java/com/exasol/sql/dml/merge/NotMatchedClause.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+ * Representation of the {@code WHEN NOT MATCHED} part of an SQL {@code MERGE statement}.
+ */
+public class NotMatchedClause extends AbstractFragment implements MergeFragment {
+    private MergeInsertClause mergeInsertClause;
+
+    public NotMatchedClause(final Fragment root) {
+        super(root);
+    }
+
+    public MergeInsertClause thenInsert() {
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/sql/dml/merge/NotMatchedClause.java]: Added.

File: src/main/java/com/exasol/sql/expression/ExpressionTerm.java [hunk 15]
```diff
@@ -0,0 +1,23 @@
+package com.exasol.sql.expression;
+
+public abstract class ExpressionTerm extends AbstractValueExpression {
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/sql/expression/ExpressionTerm.java]: Added.

File: src/main/java/com/exasol/sql/expression/rendering/ValueExpressionRenderer.java [hunk 16]
```diff
@@ -1,7 +1,11 @@
 package com.exasol.sql.expression.rendering;
 
 import com.exasol.sql.UnnamedPlaceholder;
-import com.exasol.sql.expression.*;
+import com.exasol.sql.expression.ColumnReference;
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/sql/expression/rendering/ValueExpressionRenderer.java]: Done.

File: src/main/java/com/exasol/sql/expression/rendering/ValueExpressionRenderer.java (hunk 5 above)
Comment (redcatbear) [reply src/main/java/com/exasol/sql/expression/rendering/ValueExpressionRenderer.java]: `!` is a unary operator while `!=` is a binary operator. Unary operator precedence is always obvious even without extra brackets.

File: src/test/java/com/exasol/sql/ddl/create/TestCreateTable.java (hunk 7 above)
Comment (redcatbear) [reply src/test/java/com/exasol/sql/ddl/create/TestCreateTable.java]: Fixed.

File: src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java (hunk 8 above)
Comment (redcatbear) [reply src/test/java/com/exasol/sql/ddl/drop/TestDropSchema.java]: Returns true if the statement has "IfExists",

File: src/test/java/com/exasol/sql/expression/ColumnReferenceTest.java (hunk 9 above)
Comment (redcatbear) [reply src/test/java/com/exasol/sql/expression/ColumnReferenceTest.java]: Removed.

File: doc/user_guide/statements/merge.md [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+    .thenInsert(1, "Hello");
+```
+
+Check the JavaDoc of class `Merge` for a more information.
+
+### Rendering `MERGE` Statements
+
+Use the `MergeRenderer` to render `Merge` objects into SQL strings.
+
+For a more general introduction please refer to ["Rendering SQL Statements into Strings"](rendering.md).
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply doc/user_guide/statements/merge.md]: Fixed.

**Pull request merged**